src/main/scala/chisel3/util/Counter.scala

```diff
@@ -19,27 +19,50 @@ import chisel3.internal.naming.chiselName  // can't use chisel3_ version because
   *     ...
   *   }
   * }}}
-  *
-  * @param n number of counts before the counter resets (or one more than the
-  * maximum output value of the counter), need not be a power of two
   */
 @chiselName
-class Counter(val n: Int) {
-  require(n >= 0, s"Counter value must be nonnegative, got: $n")
-  val value = if (n > 1) RegInit(0.U(log2Ceil(n).W)) else 0.U
+class Counter private (r: Range) {
```

Comment:
Related to https://github.com/freechipsproject/chisel3/pull/1363, better would probably be `NumericRange[BigInt]` (while having factory methods for both kinds of ranges). That being said, because this is private it should be easy enough to change so I think it's best to merge this PR and leave the `BigInt` form for later 🙂

Reply:
Cool, that makes sense :+1: